packages/edit-navigation/src/store/actions.js

```diff
@@ -0,0 +1,251 @@
+/**
+ * External dependencies
+ */
+import { invert, keyBy, omit } from 'lodash';
+
+/**
+ * WordPress dependencies
+ */
+import { __ } from '@wordpress/i18n';
+
+/**
+ * Internal dependencies
+ */
+import { select, resolveMenuItems, dispatch, apiFetch } from './controls';
+import { uuidv4, menuItemsQuery, KIND, POST_TYPE } from './utils';
+
+// Hits POST /wp/v2/menu-items once for every Link block that doesn't have an
+// associated menu item. (IDK what a good name for this is.)
+export const createMissingMenuItems = serializeProcessing( function* ( post ) {
+	const menuId = post.meta.menuId;
+	const mapping = post.meta.menuItemIdToClientId;
+	const clientIdToMenuId = invert( mapping );
+
+	const stack = [ post.blocks[ 0 ] ];
+	while ( stack.length ) {
+		const block = stack.pop();
+		if ( ! ( block.clientId in clientIdToMenuId ) ) {
+			const menuItem = yield apiFetch( {
+				path: `/__experimental/menu-items`,
+				method: 'POST',
+				data: {
+					title: 'Placeholder',
+					url: 'Placeholder',
+					menu_order: 0,
+				},
+			} );
+
+			mapping[ menuItem.id ] = block.clientId;
+			const menuItems = yield resolveMenuItems( menuId );
+			yield dispatch(
+				'core',
+				'receiveEntityRecords',
+				'root',
+				'menuItem',
+				[ ...menuItems, menuItem ],
+				menuItemsQuery( menuId ),
+				false
+			);
+		}
+		stack.push( ...block.innerBlocks );
+	}
+
+	yield dispatch( 'core', 'editEntityRecord', KIND, POST_TYPE, post.id, {
+		meta: {
+			...post.meta,
+			menuItemIdToClientId: mapping,
+		},
+	} );
+} );
+
+export const saveNavigationPost = serializeProcessing( function* ( post ) {
+	const menuId = post.meta.menuId;
+	const menuItemsByClientId = mapMenuItemsByClientId(
+		yield resolveMenuItems( menuId ),
+		post.meta.menuItemIdToClientId
+	);
+
+	try {
+		yield* batchSave( menuId, menuItemsByClientId, post.blocks[ 0 ] );
+		yield dispatch(
+			'core/notices',
+			'createSuccessNotice',
+			__( 'Navigation saved.' ),
+			{
+				type: 'snackbar',
+			}
+		);
+	} catch ( e ) {
+		yield dispatch(
+			'core/notices',
+			'createErrorNotice',
+			__( 'There was an error.' ),
+			{
+				type: 'snackbar',
+			}
+		);
```

Comment:
It would be nice to extract these notices outside of the save action, but that would require some way to wait for `serialProcessing()` to finish. I think it could wait until another iteration.